### DISCUSSION TITLE
Procedure or function "" has too many arguments specified on upgrading version

### DISCUSSION THREAD

**Discussion by lv1986** [Mimetis/Dotmim.Sync#848] [Q&A] [answered]
Hi, I am upgrading Dotmim.Sync to the latest version 0.9.7. Sync is working fine with the old version but syncing to Sqlite database is taking too much time so we have decided to upgrade to the latest version. but I am getting this error:

**Procedure or function "" has too many arguments specified.**

Steps I have followed:

- Created new Asp .net core Wep Application using newest version of Dotmim.
- Deprovision database
- Provision Database
- Deleted my local database
- Ran Sync method.

Here is my sample code from console application to sync data from SQL server to Sqlite database:

```
using Dotmim.Sync;
using Dotmim.Sync.Sqlite;
using Dotmim.Sync.Web.Client;

var serverOrchestrator = new WebRemoteOrchestrator("http://localhost:7845/api/sync");

// Second provider is using plain old Sql Server provider,
// relying on triggers and tracking tables to create the sync environment
string clientConnectionString = @"Data Source=C:\Users\Owner\AppData\Roaming\SyncTest.db";
SqliteSyncProvider clientProvider = new SqliteSyncProvider(clientConnectionString);
var options = new SyncOptions
{
    BatchSize = 2000
};

serverOrchestrator.AddCustomHeader("Authorization", "Bearer " + "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxNCIsImVtYWlsIjoic2x0ZXN0QHlvcG1haWwuY29tIiwianRpIjoiYTFiYmJhZmQtZDhiNC00MDBhLTkwNDgtNTFkYTFhNzExMDRlIiwiZXhwIjoxNjY2NzY1NTMyLCJpc3MiOiJTbWFydExheWVyIiwiYXVkIjoiU21hcnRMYXllciJ9.alB0eKprmSoqwIPBTqJKBltf6UidYB9gXe2CsdeWAfo");

// Creating an agent that will handle all the process
var agent = new SyncAgent(clientProvider, serverOrchestrator, options);

var parameters = new SyncParameters
{
    { "TenantId", "3" }
};
var progress = new SynchronousProgress<ProgressArgs>(
   pa => Console.WriteLine($"{pa.ProgressPercentage:p}\t {pa.Message}"));

do
{
    // Launch the sync process
    try
    {
        var s1 = await agent.SynchronizeAsync("DefaultScope", Dotmim.Sync.Enumerations.SyncType.Reinitialize, parameters, progress);
        Console.WriteLine(s1);
    }
    catch(Exception ex)
    { 
    
    }
    // Write results
   

} while (Console.ReadKey().Key != ConsoleKey.Escape);

Console.WriteLine("End");

```
Can someone please let me know what I am doing wrong.

**Comment by Mimetis** [comment, marked as answer]
I don't know what happens here, Need more details on the error
But what I'm thinking here is that your server has not been correctly upgraded.
Can you try with a fresh server database without any DMS metadata (no stored proc, no triggers no tracking tables, no scopes tables) and see if it's working correctly.?
You can call [DropAllAsync](https://dotmimsync.readthedocs.io/Orchestrators.html?highlight=DropAllAsync#dropallasync) to drop everything if you don't know how to do it manually, using a `RemoteOrchestrator` to target your server database (don't do it on a prod database !!!)

Another point: Don't call sync with a scope name set to `DefaultScope` since it's the default scope (and it can lead to potential edges issues i did not test)

**Reply by lv1986** [reply]
Thanks for your quick reply. 
I have tested with the fresh database and now I am getting this error:

```
Object cannot be cast from DBNull to other types.
-----------------------
   at Dotmim.Sync.RemoteOrchestrator.InternalApplyThenGetChangesAsync(ScopeInfoClient cScopeInfoClient, ScopeInfo cScopeInfo, SyncContext context, ClientSyncChanges clientChanges, DbConnection connection, DbTransaction transaction, CancellationToken cancellationToken, IProgress`1 progress)
   at Dotmim.Sync.Web.Server.WebServerAgent.ApplyThenGetChangesAsync2(HttpContext httpContext, HttpMessageSendChangesRequest httpMessage, SessionCache sessionCache, Int32 clientBatchSize, CancellationToken cancellationToken, IProgress`1 progress)
   at Dotmim.Sync.Web.Server.WebServerAgent.HandleRequestAsync(HttpContext httpContext, Action`1 action, CancellationToken cancellationToken, IProgress`1 progress)
-----------------------
-----------------------
INNER EXCEPTION
-----------------------
Object cannot be cast from DBNull to other types.
-----------------------
   at System.DBNull.System.IConvertible.ToInt64(IFormatProvider provider)
   at System.Convert.ToInt64(Object value)
   at Dotmim.Sync.BaseOrchestrator.InternalGetLocalTimestampAsync(SyncContext context, DbConnection connection, DbTransaction transaction, CancellationToken cancellationToken, IProgress`1 progress)
   at Dotmim.Sync.BaseOrchestrator.InternalGetLocalTimestampAsync(SyncContext context, DbConnection connection, DbTransaction transaction, CancellationToken cancellationToken, IProgress`1 progress)
   at Dotmim.Sync.RemoteOrchestrator.InternalApplyThenGetChangesAsync(ScopeInfoClient cScopeInfoClient, ScopeInfo cScopeInfo, SyncContext context, ClientSyncChanges clientChanges, DbConnection connection, DbTransaction transaction, CancellationToken cancellationToken, IProgress`1 progress)
   at Dotmim.Sync.RemoteOrchestrator.InternalApplyThenGetChangesAsync(ScopeInfoClient cScopeInfoClient, ScopeInfo cScopeInfo, SyncContext context, ClientSyncChanges clientChanges, DbConnection connection, DbTransaction transaction, CancellationToken cancellationToken, IProgress`1 progress)
   at Dotmim.Sync.RemoteOrchestrator.InternalApplyThenGetChangesAsync(ScopeInfoClient cScopeInfoClient, ScopeInfo cScopeInfo, SyncContext context, ClientSyncChanges clientChanges, DbConnection connection, DbTransaction transaction, CancellationToken cancellationToken, IProgress`1 progress)
-----------------------
```

I have used just a simple table for testing:

```
CREATE TABLE Test(
	[ID] NVARCHAR(50)  NOT NULL PRIMARY KEY,
	[CountryName] [nvarchar](100) NOT NULL
) ON [PRIMARY]

CREATE TABLE [dbo].[Country](
	[ID] [int] IDENTITY(1,1) NOT NULL PRIMARY KEY,
	[CountryName] [nvarchar](100) NOT NULL
) ON [PRIMARY]
```

 Steps: 

1.  Created new database.
2.  Run the Asp core web app
3. Run the console application 

I also noticed that there no tracking tables or stored procedures are created. If I manually provision the server procedures and tracking tables are created but same error is coming when trying to sync.

**Reply by Mimetis** [reply]
wow?
Can you share a github repo with your sample (a console application + a web api project + small script database) ?

**Reply by Mimetis** [reply]
are u using **Sql Change Tracking** ?

**Reply by lv1986** [reply]
No, I am not using SQL Change Tracking.

**Reply by Mimetis** [reply]
please create a **Github Repository** with **BOTH** **Web Api** and **Sample App**

**Reply by lv1986** [reply]
https://github.com/lv1986/Akoundal

**Reply by Mimetis** [reply]
I'm sorry, you need to understand what is a github project :)
I need you to share a project source code in Github
Please upskill on github and what is the purpose of open source projects and share your project correctly :)

**Reply by lv1986** [reply]
https://github.com/lv1986/ConsoleApp_SyncTest
https://github.com/lv1986/MVC_Core

Is it ok now

**Reply by Mimetis** [reply]
No it's not ok.

I've created a repository with your solution: https://github.com/Mimetis/LV1986

Look at it carefully, and please next time, use it as a template

- 1 single repository containing the solution file and 2 projects
- 1 we api project
- 1 console application
- 1 sql script CONTAINING EVERYTHING NEEDED TO CREATE THE DATABASE
- 1 .gitignore file to encapsulate everything

You are using `SqlSyncChangeTrackingProvider`, so **YES**, **YOU ARE USING CHANGE TRACKING.**

I'm sorry, I don't have the necessary time to invest more in your case.
The solution I provided here is working fine for me.

![image](https://user-images.githubusercontent.com/4592555/198092130-d12384b4-c5a6-4c80-b76f-010c6c5480f3.png)

![image](https://user-images.githubusercontent.com/4592555/198092423-f27c0dbc-c2c9-4a46-a594-f1b95ebd53ad.png)

![image](https://user-images.githubusercontent.com/4592555/198092520-8e72628f-39d4-4433-9552-bbe2d5de8ca8.png)


Just use the correct provider please...
And read the doc : https://dotmimsync.readthedocs.io/

**Comment by lv1986** [comment]
Sorry for the late reply and thanks a lot. After applying Dropall, then Deprovision, and Then Provision. Syncing is working fine. 
I also noticed that syncing speed is also very fast in the current version, Thanks for this. 
One issue I have faced is: In your documentation, it is written that: `DisableConstraintsOnApplyChanges` is by default true. But It is not true in my case for .Net Core Project. I have to manually set this value to true. 

One More thing I have noticed: New framework is changing DataTypes of Database columns in SQLite, and All DateTime columns are not converting into text. Do you have any idea why this is happening?

**Reply by Mimetis** [reply]
> I also noticed that syncing speed is also very fast in the current version, Thanks for this.

Thanks :)

>  it is written that: DisableConstraintsOnApplyChanges is by default true. But It is not true in my case for .Net Core Project. I have to manually set this value to true.

Oups, you're right, it's `false` by default. I will edit the doc to fix it !

>  New framework is changing DataTypes of Database columns in SQLite, and All DateTime columns are not converting into text. Do you have any idea why this is happening?

I don't understand, can you explain more (open a new issue if needed) or share a reprod sample  ?

**Reply by lv1986** [reply]
> New framework is changing DataTypes of Database columns in SQLite, and All DateTime columns are not converting into text. Do you have any idea why this is happening?
>
Sorry my mistake there is a typo "DateTime columns are not converting into text datatype" I mean to say ""DateTime columns are **now** converting into text"" 

I am starting a new discussion for this.
